File: specialized/sudoku2.nim
#!./make.py --10
#INFO: from c to nim (by gemini3)
import strutils, bitops, std/syncio

type
  Sudoku = array[81, char]
  Masks = array[9, uint16]

var
  rowMask, colMask, boxMask: Masks

proc getBox(row, col: int): int {.inline.} =
  (row div 3) * 3 + (col div 3)

proc initMasks(g: Sudoku) =
  for i in 0..<9:
    rowMask[i] = 0
    colMask[i] = 0
    boxMask[i] = 0
  
  for i in 0..<81:
    if g[i] != '.':
      let bit = 1.uint16 shl (ord(g[i]) - ord('1'))
      let row = i div 9
      let col = i mod 9
      rowMask[row] = rowMask[row] or bit
      colMask[col] = colMask[col] or bit
      boxMask[getBox(row, col)] = boxMask[getBox(row, col)] or bit

proc getCandidates(idx: int): uint16 {.inline.} =
  let row = idx div 9
  let col = idx mod 9
  return 0x1FF.uint16 and not (rowMask[row] or colMask[col] or boxMask[getBox(row, col)])

proc place(idx: int, val: char) {.inline.} =
  let bit = 1.uint16 shl (ord(val) - ord('1'))
  let row = idx div 9
  let col = idx mod 9
  rowMask[row] = rowMask[row] or bit
  colMask[col] = colMask[col] or bit
  boxMask[getBox(row, col)] = boxMask[getBox(row, col)] or bit

proc propagate(g: var Sudoku): bool =
  var changed = true
  var iterations = 0
  
  while changed and iterations < 15:
    changed = false
    iterations.inc
    
    # 1. Naked Singles
    for i in 0..<81:
      if g[i] == '.':
        let cand = getCandidates(i)
        if cand == 0: return false
        if countSetBits(cand) == 1:
          # Correction ici : utilisation de countTrailingZeroBits
          let digitIdx = countTrailingZeroBits(cand)
          g[i] = chr(ord('1') + digitIdx)
          place(i, g[i])
          changed = true

    # 2. Hidden Singles (Lignes)
    for row in 0..<9:
      for digit in 0..<9:
        let bit = 1.uint16 shl digit
        if (rowMask[row] and bit) != 0: continue
        var count = 0
        var lastCol = -1
        for col in 0..<9:
          let idx = row * 9 + col
          if g[idx] == '.' and (getCandidates(idx) and bit) != 0:
            count.inc
            lastCol = col
            if count > 1: break
        if count == 1:
          let idx = row * 9 + lastCol
          g[idx] = chr(ord('1') + digit)
          place(idx, g[idx])
          changed = true

    # 3. Hidden Singles (Colonnes)
    for col in 0..<9:
      for digit in 0..<9:
        let bit = 1.uint16 shl digit
        if (colMask[col] and bit) != 0: continue
        var count = 0
        var lastRow = -1
        for row in 0..<9:
          let idx = row * 9 + col
          if g[idx] == '.' and (getCandidates(idx) and bit) != 0:
            count.inc
            lastRow = row
            if count > 1: break
        if count == 1:
          let idx = lastRow * 9 + col
          g[idx] = chr(ord('1') + digit)
          place(idx, g[idx])
          changed = true
  
  return true

proc solve(g: var Sudoku): bool =
  if not propagate(g): return false

  var bestIdx = -1
  var minCount = 10
  var bestMask: uint16 = 0

  for i in 0..<81:
    if g[i] == '.':
      let cand = getCandidates(i)
      if cand == 0: return false
      let cnt = countSetBits(cand)
      if cnt < minCount:
        minCount = cnt
        bestIdx = i
        bestMask = cand
        if cnt == 1: break

  if bestIdx == -1: return true

  # Sauvegarde par valeur (très rapide en Nim)
  let backupG = g
  let backupRow = rowMask
  let backupCol = colMask
  let backupBox = boxMask

  for digit in 0..<9:
    if (bestMask and (1.uint16 shl digit)) != 0:
      g[bestIdx] = chr(ord('1') + digit)
      place(bestIdx, g[bestIdx])
      
      if solve(g): return true
      
      # Restauration
      g = backupG
      rowMask = backupRow
      colMask = backupCol
      boxMask = backupBox
  
  return false

proc main() =
  var line: string
  while stdin.readLine(line):
    let trimmed = line.strip()
    if trimmed.len >= 81:
      var grid: Sudoku
      for i in 0..<81: grid[i] = trimmed[i]
      initMasks(grid)
      if solve(grid):
        var res = newStringOfCap(81)
        for c in grid: res.add(c)
        echo res
      else:
        echo "No solution"

main()

File: specialized/sudoku2.go
package main
//INFO: from c to go (by gemini3)
import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
)

type Sudoku [81]byte
type Masks [9]uint16

type Solver struct {
	rowMask Masks
	colMask Masks
	boxMask Masks
}

func (s *Solver) initMasks(g *Sudoku) {
	for i := 0; i < 9; i++ {
		s.rowMask[i], s.colMask[i], s.boxMask[i] = 0, 0, 0
	}
	for i := 0; i < 81; i++ {
		if g[i] != '.' {
			bit := uint16(1) << (g[i] - '1')
			row, col := i/9, i%9
			s.rowMask[row] |= bit
			s.colMask[col] |= bit
			s.boxMask[(row/3)*3+(col/3)] |= bit
		}
	}
}

func (s *Solver) getCandidates(idx int) uint16 {
	row, col := idx/9, idx%9
	return 0x1FF & ^(s.rowMask[row] | s.colMask[col] | s.boxMask[(row/3)*3+(col/3)])
}

func (s *Solver) place(idx int, val byte) {
	bit := uint16(1) << (val - '1')
	row, col := idx/9, idx%9
	s.rowMask[row] |= bit
	s.colMask[col] |= bit
	s.boxMask[(row/3)*3+(col/3)] |= bit
}

func (s *Solver) propagate(g *Sudoku) bool {
	changed := true
	for iterations := 0; changed && iterations < 15; iterations++ {
		changed = false
		// 1. Naked Singles
		for i := 0; i < 81; i++ {
			if g[i] == '.' {
				cand := s.getCandidates(i)
				if cand == 0 { return false }
				if bits.OnesCount16(cand) == 1 {
					digitIdx := bits.TrailingZeros16(cand)
					g[i] = byte('1' + digitIdx)
					s.place(i, g[i])
					changed = true
				}
			}
		}
		// 2. Hidden Singles (Lignes)
		for row := 0; row < 9; row++ {
			for digit := uint16(0); digit < 9; digit++ {
				bit := uint16(1) << digit
				if (s.rowMask[row] & bit) != 0 { continue }
				count, lastCol := 0, -1
				for col := 0; col < 9; col++ {
					idx := row*9 + col
					if g[idx] == '.' && (s.getCandidates(idx)&bit) != 0 {
						count++
						lastCol = col
						if count > 1 { break }
					}
				}
				if count == 1 {
					idx := row*9 + lastCol
					g[idx] = byte('1' + digit)
					s.place(idx, g[idx])
					changed = true
				}
			}
		}
	}
	return true
}

func (s *Solver) solve(g *Sudoku) bool {
	if !s.propagate(g) { return false }

	bestIdx, minCount, bestMask := -1, 10, uint16(0)
	for i := 0; i < 81; i++ {
		if g[i] == '.' {
			cand := s.getCandidates(i)
			if cand == 0 { return false }
			cnt := bits.OnesCount16(cand)
			if cnt < minCount {
				minCount, bestIdx, bestMask = cnt, i, cand
				if cnt == 1 { break }
			}
		}
	}

	if bestIdx == -1 { return true }

	// Sauvegarde de l'état (Go copie les tableaux par valeur)
	backupG := *g
	backupRow := s.rowMask
	backupCol := s.colMask
	backupBox := s.boxMask

	for digit := uint16(0); digit < 9; digit++ {
		if (bestMask & (1 << digit)) != 0 {
			g[bestIdx] = byte('1' + digit)
			s.place(bestIdx, g[bestIdx])

			if s.solve(g) { return true }

			// Restauration
			*g = backupG
			s.rowMask = backupRow
			s.colMask = backupCol
			s.boxMask = backupBox
		}
	}
	return false
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	solver := &Solver{}
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) >= 81 {
			var grid Sudoku
			copy(grid[:], line[:81])
			solver.initMasks(&grid)
			if solver.solve(&grid) {
				fmt.Println(string(grid[:]))
			} else {
				fmt.Println("No solution")
			}
		}
	}
}